src/util/bson/bson.cpp

```diff
@@ -581,29 +584,30 @@ class Parser : public nlohmann::json_sax<json> {
 
         BsonContainer& operator=(BsonContainer&& v)
         {
+            if (&v == this) return *this;
             m_type = v.m_type;
             if (m_type == DOCUMENT) {
                 if (document) delete document;
                 document = v.document;
-                v.document = NULL;
+                v.document = nullptr;
             } else {
                 if (array) delete array;
                 array = v.array;
-                v.array = NULL;
+                v.array = nullptr;
             }
             return *this;
         }
 
         BsonContainer(BsonContainer&& v) {
+            if (m_type != v.m_type) this->~BsonContainer();
```

Comment:
At this point (in a constructor) none of the member variables have been initialized, so they hold random values (though on mac everything is zero-initialized). Calling the destructor here will go and delete random memory which can cause a crash (if you're lucky). However, you would want to do this type of cleanup in the above `operator=` methods because those are still leaking memory when the type changes.